src/main/java/io/vertx/core/http/impl/pool/Pool.java

```diff
@@ -0,0 +1,315 @@
+   *
+   * @param <C> the connection type
+   */
+  public static class Holder<C> {
+
+    C connection;        // The connection instance
+    long concurrency;    // How many times we can borrow from the connection
+    long capacity;       // How many times the connection is currently borrowed (0 <= capacity <= concurrency)
+    Channel channel;     // Transport channel
+    ContextImpl context; // Context associated with the connection
+    long weight;         // The weight that participates in the pool weight
+
+  }
+  private static final Logger log = LoggerFactory.getLogger(Pool.class);
+
+  private final ConnectionProvider<C> connector;
+  private final long maxWeight;
+  private final int queueMaxSize;
+
+  private final Queue<Waiter<C>> waiters = new ArrayDeque<>();
+  private final Set<Holder<C>> all;
+  private final Deque<Holder<C>> available;
+  private boolean closed;
+  private long capacity;
+  private long weight;
+  private final Handler<Void> poolClosed;
+  private final BiConsumer<Channel, C> connectionAdded;
+  private final BiConsumer<Channel, C> connectionRemoved;
+
+  public Pool(ConnectionProvider<C> connector,
+              int queueMaxSize,
+              long maxWeight,
+              Handler<Void> poolClosed,
+              BiConsumer<Channel, C> connectionAdded,
+              BiConsumer<Channel, C> connectionRemoved) {
+    this.maxWeight = maxWeight;
+    this.connector = connector;
+    this.queueMaxSize = queueMaxSize;
+    this.poolClosed = poolClosed;
+    this.all = new HashSet<>();
+    this.available = new ArrayDeque<>();
+    this.connectionAdded = connectionAdded;
+    this.connectionRemoved = connectionRemoved;
+  }
+
+  public synchronized boolean getConnection(Waiter<C> waiter) {
+    if (closed) {
+      return false;
+    }
+    // Enqueue
+    if (capacity > 0 || weight < maxWeight || (queueMaxSize < 0 || waiters.size() < queueMaxSize)) {
+      waiters.add(waiter);
+      checkPending();
+    } else {
+      waiter.handleFailure(waiter.context, new ConnectionPoolTooBusyException("Connection pool reached max wait queue size of " + queueMaxSize));
+    }
+    return true;
+  }
+
+  private void checkPending() {
+    while (true) {
+      Waiter<C> waiter = waiters.peek();
+      if (waiter == null) {
+        break;
+      }
+      if (capacity > 0) {
+        capacity--;
+        Holder<C> conn = available.peek();
+        if (--conn.capacity == 0) {
+          available.poll();
+        }
+        waiters.poll();
+        ContextImpl ctx = conn.context;
+        ctx.nettyEventLoop().execute(() -> {
+          if (connector.isValid(conn.connection)) {
+            boolean handled = deliverToWaiter(conn, waiter);
+            if (!handled) {
+              synchronized (Pool.this) {
+                recycleConnection(conn, 1,false);
+                checkPending();
+              }
+            }
+          } else {
+            synchronized (Pool.this) {
+              waiters.add(waiter);
```

Comment:
Shouldn't we check the `waiters` queue constraint here (`queueMaxSize`)?

Reply:
yes, I will rework the pool so it's better handled by removing the connection validity check that will be replaced by a simple pool eviction and avoid this check